### DISCUSSION TITLE
How To Use FileWriterTool To Write Code To File

### DISCUSSION THREAD

**Discussion by jamesbyars** [crewAIInc/crewAI#1526] [Q&A]
I'm working on a crew that translates business requirements into working software (python). What I have is an Agent that writes the code to satisfy the requirements but the agent is not able to write the code to a file.

I don't want to hardcode a specific file to write the code to because the application can be made up of multiple files (think DAOs, views, controllers, tests, etc). 

I want the crew to be able to:
* read the existing filesystem to comprehend the current state of the application
* update or write the necessary code to satisfy the requirements
* write the changes to the file system to reflect the changes

I've looked at the [LangChain FileManagementToolkit](https://python.langchain.com/docs/integrations/tools/filesystem/) and the [CrewAI Tool FileWriterTool](https://docs.crewai.com/tools/filewritetool). 

I'm unable to get either tool working to write the task generated output to a file using the agent.  Sample code (incomplete) below.

```python

from crewai import LLM, Agent, Task, Crew
from langchain.agents.agent_toolkits import FileManagementToolkit
from crewai_tools.tools.file_writer_tool.file_writer_tool import FileWriterToolInput

llm = LLM(model="bedrock/anthropic.claude-3-haiku-20240307-v1:0")

filesystem_tools = FileManagementToolkit(
    root_dir='/Users/[dir]/code/crewai_poc/src',
).get_tools()

software_engineer = Agent(
    role='Senior Python Engineer',
    goal='''Design and implement solutions based on user stories and requirements as defined. Deliver high quality code with an emphasis on SOLID software engineering techniques''',
    backstory='''You have worked in high pressure environments your entire career. You understand how to use Python to meet deadlines and delivery high quality, working software.''',
    verbose=True,
    allow_delegation=False,
    llm=llm,
    tool=[
        filesystem_tools,
    ],
)

create_test = Task(
    description='''
    YOU MUST USE the tool to write an updated 
      version of the test to the file 
      system in the following path: '.' 
      replacing the text content with the suggestions 
      provided.
      
      You first write the file then your final answer 
      MUST be the updated test content.

      RULES
      -----
      - NEVER WRITE \\n (newlines as string) on the file, just the code.
      - Keep the same component imports and don't use new components.
      - NEVER USE COMPONENTS THAT ARE NOT IMPORTED.
      - ALL COMPONENTS USED SHOULD BE IMPORTED, don't make up components.
      - Save the file as with `.py` extension.

      If you follow the rules I'll give you a $100 tip!!! 
      MY LIFE DEPEND ON YOU FOLLOWING IT!  
    
    Using the requirements provided below, implement unit tests in python using the unittest module.
    
    {requirements}''',
    expected_output='''A file containing the test code''',
    agent=software_engineer,
    verbose=True,
    output_pydantic=FileWriterToolInput
)

crew = Crew(
    agents=[software_engineer],
    tasks=[
        create_test,
    ],
    verbose=True,
)

inputs = {
    'requirements': 'hello world application that prints out "Hello World!"',
}

result = crew.kickoff(inputs=inputs)

```

**Comment by Hassanmahmood4** [comment]
A couple of things stand out to me.

First, I think there's a mismatch between the tools you're providing to the agent and what you're expecting it to do.

You're creating:

```python
filesystem_tools = FileManagementToolkit(
    root_dir="/Users/[dir]/code/crewai_poc/src",
).get_tools()
```

but then passing:

```python
tool=[
    filesystem_tools,
],
```

Shouldn't that be:

```python
tools=filesystem_tools,
```

instead?

`get_tools()` already returns a list of tools. Wrapping it in another list may prevent CrewAI from exposing the individual tools correctly to the agent.

Second, `output_pydantic=FileWriterToolInput` only validates the final output structure. It doesn't actually invoke the file-writing tool. The agent still needs to explicitly call a tool during execution.

I'd verify that the agent actually sees the available tools by checking the verbose output. Does it ever attempt something like `write_file` or `write`?

Also, if your goal is:

- inspect the existing project structure,
- decide which files need to be created/modified,
- read existing files,
- write updated files,

you may want to model this as a sequence of tasks instead of a single task:

1. Explore the filesystem (`list_directory`, `read_file`)
2. Plan the required changes
3. Execute the writes (`write_file`)
4. Return a summary of what changed

Trying to generate code and persist it to multiple files in one step can make tool selection harder for the agent.

A few questions that would help narrow this down:

- Which CrewAI version are you using?
- When running with `verbose=True`, do you actually see the agent attempting to invoke any file tools?
- If you print `filesystem_tools`, do you see the expected `read_file`, `write_file`, `list_directory`, etc. tools exposed to the agent?

My first suspicion would be the nested list:

```python
tool=[filesystem_tools]
```

instead of:

```python
tools=filesystem_tools
```

which could prevent the file tools from being registered correctly.
